rudder-core/src/main/scala/com/normation/rudder/repository/jdbc/ExpectedReportsJdbcRepository.scala

```diff
@@ -295,10 +434,67 @@ class FindExpectedReportsJdbcRepository(
                    jdbcTemplate.query(expectedReportsQuery, params, ReportAndNodeMapper).asScala
                  } }
     } yield {
-      toExpectedReports(entries)
+      mapEntries(entries)
     }
   }
 
+
+  /**
+   * Group a list of reports by node and rule
+   * On the resulting map, we only have nodeId for nodeId with actual node config.
+   * The returned nodeConfigId is the most recent available.
+   */
+  private[this] def toNodeExpectedReports(nodeConfigIds: Set[NodeAndConfigId], entries:Seq[ReportAndNodeMapping]) : Map[NodeId, Map[NodeConfigId, Map[SerialedRuleId, RuleNodeExpectedReports]]] = {
+
+    val t0 = System.currentTimeMillis
+
+    val nodeIds = nodeConfigIds.groupBy { _.nodeId }.mapValues { _.toSeq.map(_.version) }
+    val reportsForNode = entries.groupBy { _.nodeId }
+
+    val t1 = System.currentTimeMillis
+    TimingDebugLogger.debug(s"GetExpectedReports: toNodeExpectedReports: groupBy: ${t1-t0}ms")
+
+    var t_byConfigId = 0l
+    var n = 0
+
+    val res = nodeIds.map { case (nodeId, configIds) =>
+      val tx_0 = System.currentTimeMillis
+
+      val byConfigId = configIds.map { configId =>
+        val reportsForId = reportsForNode.getOrElse(nodeId, Seq()).filter { _.nodeConfigVersions.exists { _ == configId } }
```

Comment:
I know that i'm a bit slow today, but a comment on what this algorithm do would help the future maintainer